### ISSUE TITLE
Semantic break introduced by VBOperationFactory refactoring

### ISSUE THREAD

**Issue by mavasani** [dotnet/roslyn#19344]
See code in master: http://source.roslyn.io/#Microsoft.CodeAnalysis.VisualBasic/BoundTree/Expression.vb,1467

The operation kind depends on the length of the arguments in master, but is not so in features/operation branch: https://github.com/dotnet/roslyn/blob/features/ioperation/src/Compilers/Core/Portable/Generated/Operations.xml.Generated.cs#L1917

**Comment by mavasani** [comment]
@heejaechang This is causing failures in some new tests in master when I attempt to merge master into features/ioperation. I am going to skip those tests with this bug, please re-enable when you fix this issue.

**Comment by mavasani** [comment]
```
Test 'Microsoft.CodeAnalysis.VisualBasic.UnitTests.Semantics.IOperationTests.ParameterReference_QueryExpression' failed: Expected and actual were different.
Expected: IOperation:  (OperationKind.None) (Syntax: 'From cust I ... t cust.Name')
  Children(1): IOperation:  (OperationKind.None) (Syntax: 'Select cust.Name')
      Children(1): IInvocationExpression ( Function System.Collections.Generic.IEnumerable(Of Customer).Select(Of System.String)(selector As System.Func(Of Customer, System.String)) As System.Collections.Generic.IEnumerable(Of System.String)) (OperationKind.InvocationExpression, Type: System.Collections.Generic.IEnumerable(Of System.String)) (Syntax: 'Select cust.Name')
          Instance Receiver: IConversionExpression (ConversionKind.Basic, Implicit) (OperationKind.ConversionExpression, Type: System.Collections.Generic.IEnumerable(Of Customer)) (Syntax: 'cust In customers')
              IOperation:  (OperationKind.None) (Syntax: 'cust In customers')
                Children(1): IOperation:  (OperationKind.None) (Syntax: 'customers')
                    Children(1): IParameterReferenceExpression: customers (OperationKind.ParameterReferenceExpression, Type: System.Collections.Generic.List(Of Customer)) (Syntax: 'customers')
          Arguments(1): IArgument (ArgumentKind.DefaultValue, Matching Parameter: selector) (OperationKind.Argument) (Syntax: 'cust.Name')
              IConversionExpression (ConversionKind.Basic, Implicit) (OperationKind.ConversionExpression, Type: System.Func(Of Customer, System.String)) (Syntax: 'cust.Name')
                IOperation:  (OperationKind.None) (Syntax: 'cust.Name')
                  Children(1): IOperation:  (OperationKind.None) (Syntax: 'cust.Name')
                      Children(1): IIndexedPropertyReferenceExpression: Property Customer.Name As System.String (OperationKind.PropertyReferenceExpression, Type: System.String) (Syntax: 'cust.Name')
                          Instance Receiver: IOperation:  (OperationKind.None) (Syntax: 'cust')
Actual:   IOperation:  (OperationKind.None) (Syntax: 'From cust I ... t cust.Name')
  Children(1): IOperation:  (OperationKind.None) (Syntax: 'Select cust.Name')
      Children(1): IInvocationExpression ( Function System.Collections.Generic.IEnumerable(Of Customer).Select(Of System.String)(selector As System.Func(Of Customer, System.String)) As System.Collections.Generic.IEnumerable(Of System.String)) (OperationKind.InvocationExpression, Type: System.Collections.Generic.IEnumerable(Of System.String)) (Syntax: 'Select cust.Name')
          Instance Receiver: IConversionExpression (ConversionKind.Basic, Implicit) (OperationKind.ConversionExpression, Type: System.Collections.Generic.IEnumerable(Of Customer)) (Syntax: 'cust In customers')
              IOperation:  (OperationKind.None) (Syntax: 'cust In customers')
                Children(1): IOperation:  (OperationKind.None) (Syntax: 'customers')
                    Children(1): IParameterReferenceExpression: customers (OperationKind.ParameterReferenceExpression, Type: System.Collections.Generic.List(Of Customer)) (Syntax: 'customers')
          Arguments(1): IArgument (ArgumentKind.DefaultValue, Matching Parameter: selector) (OperationKind.Argument) (Syntax: 'cust.Name')
              IConversionExpression (ConversionKind.Basic, Implicit) (OperationKind.ConversionExpression, Type: System.Func(Of Customer, System.String)) (Syntax: 'cust.Name')
                IOperation:  (OperationKind.None) (Syntax: 'cust.Name')
                  Children(1): IOperation:  (OperationKind.None) (Syntax: 'cust.Name')
                      Children(1): IIndexedPropertyReferenceExpression: Property Customer.Name As System.String (OperationKind.IndexedPropertyReferenceExpression, Type: System.String) (Syntax: 'cust.Name')
                          Instance Receiver: IOperation:  (OperationKind.None) (Syntax: 'cust')
```

See the difference in IIndexedPropertyReferenceExpression dump.

**Comment by mavasani** [comment]
Ah, I also note that even though VisualBasicOperationFactory is in features/ioperation, we are still using bound nodes for IOperation: https://github.com/dotnet/roslyn/blob/features/ioperation/src/Compilers/VisualBasic/Portable/Binding/MemberSemanticModel.vb#L803

**Comment by mavasani** [comment]
@heejaechang When are you planning to move VB operations to operation factory?

**Comment by mavasani** [comment]
@heejaechang I am going to create a new file Expressions_new.vb and Statements_new.vb with my merge (this will only have BoundNodes implementing operation with children stuff). Once you delete Expressions.vb and Statements.vb (i.e. move VB to operation factory), you can just rename these files back to Expressions.vb and Statements.vb

**Comment by heejaechang** [comment]
@mavasani thank you for the finding the issue. the second one, I thought I already changed and deleted all those.. not sure why those are still there. let me check.

**Comment by mavasani** [comment]
@heejaechang I am actually confused about the VB factory status here, let me drop by..

**Comment by mavasani** [comment]
There are few failures due to cast exception at below line the VB factory in `CreateBoundParameterEqualsValueOperation`:

`Dim isInvalid As Boolean = DirectCast(boundParameterEqualsValue.Value, IOperation).IsInvalid`

I am going to skip them with this bug.

**Comment by mavasani** [comment]
`IOperationTests.SimpleCompoundAssignment` is also failing with the below:

```
Test 'Microsoft.CodeAnalysis.VisualBasic.UnitTests.Semantics.IOperationTests.SimpleCompoundAssignment' failed: Assert.Equal() Failure
Expected: AssignmentExpression
Actual:   CompoundAssignmentExpression
	IOperation\IOperationTests.vb(200,0): at Microsoft.CodeAnalysis.VisualBasic.UnitTests.Semantics.IOperationTests.SimpleCompoundAssignment()
```

I am going to skip it with this bug.

**Comment by mavasani** [comment]
@heejaechang I have merged the conflict resolution PR and skipped a bunch of bugs in VB with this bug ID. Please revert the skipped tag when fixing this. Thanks.

**Comment by heejaechang** [comment]
Will do.


  *   heejae

**Comment by jinujoseph** [comment]
closed via #19406
